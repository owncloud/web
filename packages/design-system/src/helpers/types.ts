export interface ContextualHelperDataListItem {
  text: string
  headline?: boolean
}

export interface ContextualHelperData {
  title?: string
  text?: string
  list?: ContextualHelperDataListItem[]
  readMoreLink?: string
}

export interface ContextualHelper {
  isEnabled: boolean
  data: ContextualHelperData
}

export interface PasswordPolicyRule {
  code: string
  message: string
  helperMessage?: string
  format: (number | string)[]
  verified: boolean
}

export interface PasswordPolicy {
  rules: unknown[]

  check(password: string): boolean

  missing(password: string): {
    rules: PasswordPolicyRule[]
  }
}
