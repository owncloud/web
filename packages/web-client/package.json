{
  "name": "@ownclouders/web-client",
  "version": "11.3.1",
  "description": "ownCloud web client",
  "license": "AGPL-3.0",
  "private": false,
  "type": "module",
  "author": "ownCloud GmbH <devops@owncloud.com>",
  "homepage": "https://github.com/owncloud/web/tree/master/packages/web-client",
  "repository": {
    "type": "git",
    "url": "https://github.com/owncloud/web",
    "directory": "packages/web-client"
  },
  "files": [
    "dist/"
  ],
  "publishConfig": {
    "directory": "package",
    "linkDirectory": false,
    "exports": {
      ".": {
        "default": "./dist/web-client.js",
        "require": "./dist/web-client.cjs",
        "types": "./dist/src/index.d.ts"
      },
      "./graph": {
        "default": "./dist/web-client/graph.js",
        "require": "./dist/web-client/graph.cjs",
        "types": "./dist/src/graph/index.d.ts"
      },
      "./graph/generated": {
        "default": "./dist/web-client/graph/generated.js",
        "require": "./dist/web-client/graph/generated.cjs",
        "types": "./dist/src/graph/generated/index.d.ts"
      },
      "./ocs": {
        "default": "./dist/web-client/ocs.js",
        "require": "./dist/web-client/ocs.cjs",
        "types": "./dist/src/ocs/index.d.ts"
      },
      "./sse": {
        "default": "./dist/web-client/sse.js",
        "require": "./dist/web-client/sse.cjs",
        "types": "./dist/src/sse/index.d.ts"
      },
      "./webdav": {
        "default": "./dist/web-client/webdav.js",
        "require": "./dist/web-client/webdav.cjs",
        "types": "./dist/src/webdav/index.d.ts"
      }
    }
  },
  "exports": {
    ".": {
      "default": "./src/index.ts"
    },
    "./graph": {
      "default": "./src/graph/index.ts"
    },
    "./graph/generated": {
      "default": "./src/graph/generated/index.ts"
    },
    "./ocs": {
      "default": "./src/ocs/index.ts"
    },
    "./sse": {
      "default": "./src/sse/index.ts"
    },
    "./webdav": {
      "default": "./src/webdav/index.ts"
    },
    "./l10n": {
      "default": "./l10n/translations.json"
    }
  },
  "scripts": {
    "generate-openapi": "rm -rf src/graph/generated && docker run --rm -v \"${PWD}/src/graph:/local\" openapitools/openapi-generator-cli generate -i https://raw.githubusercontent.com/owncloud/libre-graph-api/main/api/openapi-spec/v1.0.yaml  -g typescript-axios -o /local/generated",
    "vite": "vite",
    "prepublishOnly": "rm -rf ./package && clean-publish && rm -rf package/dist/tests && find package && cat package/package.json",
    "postpublish": "rm -rf ./package"
  },
  "dependencies": {
    "@casl/ability": "^6.7.1",
    "@microsoft/fetch-event-source": "^2.0.1",
    "@ownclouders/web-pkg": "workspace:*",
    "axios": "^1.7.7",
    "fast-xml-parser": "^4.5.0",
    "lodash-es": "^4.17.21",
    "luxon": "^3.5.0",
    "uuid": "^11.0.0",
    "webdav": "^5.7.1",
    "xml-js": "^1.6.11",
    "zod": "^3.23.8"
  },
  "devDependencies": {
    "@types/luxon": "3.4.2",
    "@vitejs/plugin-vue": "5.1.4",
    "clean-publish": "5.0.0",
    "vite-plugin-dts": "4.2.4",
    "vite-plugin-node-polyfills": "0.22.0",
    "vite": "5.4.8"
  }
}
