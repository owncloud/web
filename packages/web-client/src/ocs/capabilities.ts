import { AxiosInstance } from 'axios'
import get from 'lodash-es/get'

/* eslint-disable camelcase */
export interface AppProviderCapability {
  apps_url?: string
  enabled?: boolean
  new_url?: string
  open_url?: string
  version?: string
}

export interface PasswordPolicyCapability {
  min_characters?: number
  max_characters?: number
  min_lowercase_characters?: number
  min_uppercase_characters?: number
  min_digits?: number
  min_special_characters?: number
}

export interface PasswordEnforcedForCapability {
  read_only?: boolean
  read_write?: boolean
  upload_only?: boolean
  read_write_delete?: boolean
}

export interface SearchPropertyCapability {
  enabled?: boolean
}
export interface LastModifiedFilterCapability extends SearchPropertyCapability {
  keywords?: string[]
}

export interface MediaTypeCapability extends SearchPropertyCapability {
  keywords?: string[]
}

/**
 * Archiver struct within the capabilities as defined in reva
 * @see https://github.com/cs3org/reva/blob/41d5a6858c2200a61736d2c165e551b9785000d1/internal/http/services/owncloud/ocs/data/capabilities.go#L105
 */
export interface ArchiverCapability {
  enabled?: boolean
  version?: string // version is just a major version, e.g. `v2`
  formats?: string[]
  // eslint-disable-next-line camelcase
  archiver_url?: string
  max_num_files?: string
  max_size?: string
}

export interface Capabilities {
  capabilities: {
    checksums?: {
      preferredUploadType?: string
      supportedTypes?: string[]
    }
    password_policy?: PasswordPolicyCapability
    search?: {
      property?: {
        content?: SearchPropertyCapability
        mediatype?: MediaTypeCapability
        mtime?: LastModifiedFilterCapability
        name?: MediaTypeCapability
        scope?: MediaTypeCapability
        size?: MediaTypeCapability
        tag?: MediaTypeCapability
        tags?: MediaTypeCapability
        type?: MediaTypeCapability
      }
    }
    notifications?: {
      'ocs-endpoints'?: string[]
    }
    core: {
      pollinterval?: number
      status?: {
        edition?: string
        installed?: boolean
        maintenance?: boolean
        needsDbUpgrade?: boolean
        product?: string
        productname?: string
        productversion?: string
        version?: string
        versionstring?: string
      }
      'support-sse'?: boolean
      'support-url-signing'?: boolean
      'webdav-root'?: string
    }
    dav: {
      chunking?: string
      chunkingParallelUploadDisabled?: boolean
      reports?: string[]
      trashbin?: string
    }
    files: {
      app_providers?: AppProviderCapability[]
      archivers?: ArchiverCapability[]
      favorites?: boolean
      full_text_search?: boolean
      permanent_deletion?: boolean
      privateLinks?: boolean
      tags?: boolean
      tus_support?: {
        extension?: string
        http_method_override?: boolean
        max_chunk_size?: number
        resumable?: string
        version?: string
      }
      undelete?: boolean
      versioning?: boolean
      thumbnail?: {
        enabled?: boolean
        version?: string
        supportedMimeTypes?: string[]
      }
    }
    files_sharing: {
      allow_custom?: boolean
      api_enabled?: boolean
      can_rename?: boolean
      default_permissions?: number
      deny_access?: boolean
      federation?: {
        incoming?: boolean
        outgoing?: boolean
      }
      group_sharing?: boolean
      public?: {
        alias?: boolean
        can_contribute?: boolean
        can_edit?: boolean
        default_permissions?: number
        enabled?: boolean
        multiple?: boolean
        password?: {
          enforced?: boolean
          enforced_for?: PasswordEnforcedForCapability
        }
        send_mail?: boolean
        supports_upload_only?: boolean
        upload?: boolean
      }
      search_min_length?: number
      user?: {
        profile_picture?: boolean
        send_mail?: boolean
        settings?: {
          enabled?: boolean
          version?: string
        }[]
      }
      quick_link?: {
        default_role?: string
      }
    }
    spaces?: {
      enabled?: boolean
      max_quota?: number
      projects?: boolean
      version?: string
    }
    graph?: {
      'personal-data-export'?: boolean
      users: {
        change_password_self_disabled?: boolean
        create_disabled?: boolean
        delete_disabled?: boolean
        read_only_attributes?: string[]
      }
    }
  }
  version: {
    edition?: string
    major?: string
    minor?: string
    micro?: string
    product?: string
    productversion?: string
    string?: string
  }
}
/* eslint-enable camelcase */

export const GetCapabilitiesFactory = (baseURI: string, axios: AxiosInstance) => {
  const url = new URL(baseURI)
  url.pathname = [...url.pathname.split('/'), 'cloud', 'capabilities'].filter(Boolean).join('/')
  url.searchParams.append('format', 'json')
  const endpoint = url.href
  return {
    async getCapabilities(): Promise<Capabilities> {
      const response = await axios.get(endpoint)
      return get(response, 'data.ocs.data', { capabilities: null, version: null })
    }
  }
}
