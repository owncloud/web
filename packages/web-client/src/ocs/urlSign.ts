import { AxiosInstance } from 'axios'
import { urlJoin } from '../utils'
import convert from 'xml-js'
import { pbkdf2Sync } from 'crypto'

export interface UrlSignOptions {
  axiosClient: AxiosInstance
  baseURI: string
}

export type SignUrlPayload = {
  url: string
  username: string
  publicToken?: string
  publicLinkPassword?: string
}

export class UrlSign {
  private axiosClient: AxiosInstance
  private baseURI: string

  private signingKey: string

  private ALGORITHM = 'sha512'
  private TTL = 1200
  private HASH_LENGTH = 32
  private ITERATION_COUNT = 10000

  constructor({ axiosClient, baseURI }: UrlSignOptions) {
    this.axiosClient = axiosClient
    this.baseURI = baseURI
  }

  public async signUrl({ url, username, publicToken, publicLinkPassword }: SignUrlPayload) {
    const signedUrl = new URL(url)
    signedUrl.searchParams.set('OC-Credential', username)
    signedUrl.searchParams.set('OC-Date', new Date().toISOString())
    signedUrl.searchParams.set('OC-Expires', this.TTL.toString())
    signedUrl.searchParams.set('OC-Verb', 'GET')

    const hashedKey = await this.createHashedKey(
      signedUrl.toString(),
      publicToken,
      publicLinkPassword
    )

    signedUrl.searchParams.set('OC-Algo', `PBKDF2/${this.ITERATION_COUNT}-SHA512`)
    signedUrl.searchParams.set('OC-Signature', hashedKey)

    return signedUrl.toString()
  }

  private async getSignKey(publicToken?: string, publicLinkPassword?: string) {
    if (this.signingKey) {
      return this.signingKey
    }

    const data = await this.axiosClient.get(
      urlJoin(this.baseURI, 'ocs/v1.php/cloud/user/signing-key'),
      {
        params: {
          ...(publicToken && { 'public-token': publicToken })
        },
        headers: {
          'Content-Type': 'application/x-www-form-urlencoded',
          ...(publicLinkPassword && {
            Authorization: `Basic ${Buffer.from(['public', publicLinkPassword].join(':')).toString('base64')}`
          })
        }
      }
    )

    const parsedXML = convert.xml2js(data.data, { compact: true }) as any
    this.signingKey = parsedXML.ocs.data['signing-key']._text
    return this.signingKey
  }

  private async createHashedKey(url: string, publicToken?: string, publicLinkPassword?: string) {
    const signignKey = await this.getSignKey(publicToken, publicLinkPassword)
    const hashedKey = pbkdf2Sync(
      url,
      signignKey,
      this.ITERATION_COUNT,
      this.HASH_LENGTH,
      this.ALGORITHM
    )

    return hashedKey.toString('hex')
  }
}
