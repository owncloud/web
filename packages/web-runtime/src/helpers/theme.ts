import { isEqual } from 'lodash-es'
import defaultTheme from '../../themes/owncloud/theme.json'
import { v4 as uuidV4 } from 'uuid'
import { ThemingConfig } from '@ownclouders/web-pkg'

export const loadTheme = async (location = '') => {
  const defaultOwnCloudTheme = {
    defaults: {
      ...defaultTheme.clients.web.defaults,
      common: defaultTheme.common
    },
    themes: defaultTheme.clients.web.themes
  }

  if (location.split('.').pop() !== 'json') {
    if (isEqual(process.env.NODE_ENV, 'development')) {
      console.info(`Theme '${location}' does not specify a json file, using default theme.`)
    }
    return defaultOwnCloudTheme
  }

  try {
    const response = await fetch(location, { headers: { 'X-Request-ID': uuidV4() } })
    if (!response.ok) {
      console.error(`Failed to load theme '${location}', invalid response. Using default theme.`)
      return defaultOwnCloudTheme
    }

    const theme = await response.json()

    try {
      const parsedTheme = ThemingConfig.parse(theme)

      return {
        defaults: {
          common: parsedTheme.common,
          ...parsedTheme.clients.web.defaults
        },
        themes: parsedTheme.clients.web.themes
      }
    } catch (error) {
      console.error(
        `Failed to load theme '${location}', invalid theme. Using default theme.`,
        error
      )
      return defaultOwnCloudTheme
    }
  } catch {
    console.error(`Failed to load theme '${location}', using default theme.`)
    return defaultOwnCloudTheme
  }
}
