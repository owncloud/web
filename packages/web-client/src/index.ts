import { graph } from './graph'
import { ocs } from './ocs'
import { webdav } from './webdav'

export * from './errors'
export * from './helpers'
export * from './utils'
export * from './constants'

export { graph, ocs, webdav }
