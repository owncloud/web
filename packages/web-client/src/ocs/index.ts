import { Capabilities, GetCapabilitiesFactory } from './capabilities'
import { AxiosInstance } from 'axios'
import { SignUrlPayload, UrlSign } from './urlSign'

export * from './capabilities'

export interface OCS {
  getCapabilities: () => Promise<Capabilities>
  signUrl: (payload: SignUrlPayload) => Promise<string>
}

export const ocs = (baseURI: string, axiosClient: AxiosInstance): OCS => {
  const url = new URL(baseURI)
  url.pathname = [...url.pathname.split('/'), 'ocs', 'v1.php'].filter(Boolean).join('/')
  const ocsV1BaseURI = url.href

  const capabilitiesFactory = GetCapabilitiesFactory(ocsV1BaseURI, axiosClient)

  const urlSign = new UrlSign({ baseURI, axiosClient })

  return {
    getCapabilities: () => {
      return capabilitiesFactory.getCapabilities()
    },
    signUrl: (payload: SignUrlPayload) => {
      return urlSign.signUrl(payload)
    }
  }
}
