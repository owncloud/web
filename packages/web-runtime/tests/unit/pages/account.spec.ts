import account from '../../../src/pages/account.vue'
import {
  defaultComponentMocks,
  defaultPlugins,
  mockAxiosResolve,
  mount,
  mockAxiosReject
} from 'web-test-helpers'
import { mock } from 'vitest-mock-extended'
import {
  Extension,
  ExtensionPoint,
  OptionsConfig,
  useExtensionRegistry,
  useMessages,
  useResourcesStore
} from '@ownclouders/web-pkg'
import { LanguageOption, SettingsBundle, SettingsValue } from '../../../src/helpers/settings'
import { User } from '@ownclouders/web-client/graph/generated'
import { VueWrapper } from '@vue/test-utils'
import { SpaceResource } from '@ownclouders/web-client'
import { Capabilities } from '@ownclouders/web-client/ocs'

const $route = {
  meta: {
    title: 'Some Title'
  }
}

const selectors = {
  pageTitle: '.oc-page-title',
  loaderStub: 'oc-spinner-stub',
  editUrlButton: '[data-testid="account-page-edit-url-btn"]',
  editPasswordButton: '[data-testid="account-page-edit-password-btn"]',
  logoutButton: '[data-testid="account-page-logout-url-btn"]',
  accountPageInfo: '.account-page-info',
  groupNames: '[data-testid="group-names"]',
  groupNamesEmpty: '[data-testid="group-names-empty"]',
  gdprExport: '[data-testid="gdpr-export"]',
  extensionsSection: '.account-page-extensions'
}

describe('account page', () => {
  describe('public link context', () => {
    it('should render a limited view', async () => {
      const { wrapper } = getWrapper({ isUserContext: false, isPublicLinkContext: true })
      await blockLoadingState(wrapper)

      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  describe('header section', () => {
    describe('edit url button', () => {
      it('should be displayed if defined via config', async () => {
        const { wrapper } = getWrapper({
          accountEditLink: { href: '/' }
        })
        await blockLoadingState(wrapper)

        const editUrlButton = wrapper.find(selectors.editUrlButton)
        expect(editUrlButton.html()).toMatchSnapshot()
      })
      it('should not be displayed if not defined via config', async () => {
        const { wrapper } = getWrapper()
        await blockLoadingState(wrapper)

        const editUrlButton = wrapper.find(selectors.editUrlButton)
        expect(editUrlButton.exists()).toBeFalsy()
      })
    })
  })

  describe('account information section', () => {
    it('displays basic user information', async () => {
      const { wrapper } = getWrapper({
        user: mock<User>({
          onPremisesSamAccountName: 'some-username',
          displayName: 'some-displayname',
          mail: 'some-email',
          memberOf: []
        })
      })
      await blockLoadingState(wrapper)

      const accountPageInfo = wrapper.find(selectors.accountPageInfo)
      expect(accountPageInfo.html()).toMatchSnapshot()
    })

    describe('group membership', () => {
      it('displays message if not member of any groups', async () => {
        const { wrapper } = getWrapper()
        await blockLoadingState(wrapper)

        const groupNamesEmpty = wrapper.find(selectors.groupNamesEmpty)
        expect(groupNamesEmpty.exists()).toBeTruthy()
      })
      it('displays group names', async () => {
        const { wrapper } = getWrapper({
          user: mock<User>({
            memberOf: [{ displayName: 'one' }, { displayName: 'two' }, { displayName: 'three' }]
          })
        })
        await blockLoadingState(wrapper)

        const groupNames = wrapper.find(selectors.groupNames)
        expect(groupNames.html()).toMatchSnapshot()
      })
    })

    describe('Logout from all devices link', () => {
      it('should render the logout from active devices if logoutUrl is provided', async () => {
        const { wrapper } = getWrapper()
        await blockLoadingState(wrapper)

        expect(wrapper.find('[data-testid="logout"]').exists()).toBe(true)
      })
      it("shouldn't render the logout from active devices if logoutUrl isn't provided", async () => {
        const { wrapper } = getWrapper()
        await blockLoadingState(wrapper)

        wrapper.vm.logoutUrl = undefined
        expect(wrapper.find('[data-testid="logout"]').exists()).toBe(true)
      })
      it('should use url from configuration manager', async () => {
        const { wrapper } = getWrapper()
        await blockLoadingState(wrapper)

        const logoutButton = wrapper.find(selectors.logoutButton)
        expect(logoutButton.attributes('href')).toBe('https://account-manager/logout')
      })
    })
  })

  describe('Preferences section', () => {
    describe('change password button', () => {
      it('should be displayed if not disabled via capability', async () => {
        const { wrapper } = getWrapper({
          capabilities: { graph: { users: { change_password_self_disabled: false } } }
        })
        await blockLoadingState(wrapper)

        const editPasswordButton = wrapper.find(selectors.editPasswordButton)
        expect(editPasswordButton.exists()).toBeTruthy()
      })
      it('should not be displayed if disabled via capability', async () => {
        const { wrapper } = getWrapper({
          capabilities: { graph: { users: { change_password_self_disabled: true } } }
        })
        await blockLoadingState(wrapper)

        const editPasswordButton = wrapper.find(selectors.editPasswordButton)
        expect(editPasswordButton.exists()).toBeFalsy()
      })
    })
  })

  describe('Method "updateDisableEmailNotifications', () => {
    it('should show a message on success', async () => {
      const { wrapper, mocks } = getWrapper()
      await blockLoadingState(wrapper)

      mocks.$clientService.httpAuthenticated.post.mockResolvedValueOnce(mockAxiosResolve({}))
      await wrapper.vm.updateDisableEmailNotifications(true)
      const { showMessage } = useMessages()
      expect(showMessage).toHaveBeenCalled()
    })
    it('should show a message on error', async () => {
      vi.spyOn(console, 'error').mockImplementation(() => undefined)

      const { wrapper, mocks } = getWrapper()
      await blockLoadingState(wrapper)

      mocks.$clientService.httpAuthenticated.post.mockImplementation(() => mockAxiosReject('err'))
      await wrapper.vm.updateDisableEmailNotifications(true)
      const { showErrorMessage } = useMessages()
      expect(showErrorMessage).toHaveBeenCalled()
    })
  })

  describe('Method "updateSelectedLanguage', () => {
    it('should show a message on success', async () => {
      const { wrapper, mocks } = getWrapper({})
      await blockLoadingState(wrapper)

      mocks.$clientService.graphAuthenticated.users.editMe.mockResolvedValueOnce(undefined)
      await wrapper.vm.updateSelectedLanguage({ value: 'en' } as LanguageOption)
      const { showMessage } = useMessages()
      expect(showMessage).toHaveBeenCalled()
    })
    it('should show a message on error', async () => {
      vi.spyOn(console, 'error').mockImplementation(() => undefined)

      const { wrapper, mocks } = getWrapper({})
      await blockLoadingState(wrapper)

      mocks.$clientService.graphAuthenticated.users.editMe.mockRejectedValue(new Error())
      await wrapper.vm.updateSelectedLanguage({ value: 'en' } as LanguageOption)
      const { showErrorMessage } = useMessages()
      expect(showErrorMessage).toHaveBeenCalled()
    })
  })

  describe('Method "updateViewOptionsWebDavDetails', () => {
    it('should show a message on success', async () => {
      const { wrapper } = getWrapper({})
      await blockLoadingState(wrapper)

      await wrapper.vm.updateViewOptionsWebDavDetails(true)
      const { showMessage } = useMessages()
      expect(showMessage).toHaveBeenCalled()

      const { setAreWebDavDetailsShown } = useResourcesStore()
      expect(setAreWebDavDetailsShown).toHaveBeenCalled()
    })
  })

  describe('Extensions section', () => {
    it('should be hidden if no extension points offer preferences', async () => {
      const { wrapper } = getWrapper({})
      await blockLoadingState(wrapper)

      expect(wrapper.find(selectors.extensionsSection).exists()).toBeFalsy()
    })

    it('should be hidden if an extension point only has 1 or less extensions', async () => {
      const extensionPointMock = mock<ExtensionPoint<Extension>>({
        userPreference: {
          label: 'example-extension-point'
        }
      })
      const { wrapper } = getWrapper({
        extensionPoints: [extensionPointMock]
      })
      await blockLoadingState(wrapper)

      expect(wrapper.find(selectors.extensionsSection).exists()).toBeFalsy()
    })

    it('should be visible if an extension point has at least 2 extensions', async () => {
      const extensionPoint = mock<ExtensionPoint<Extension>>({
        id: 'test-extension-point',
        multiple: false,
        defaultExtensionId: 'foo-2',
        userPreference: {
          label: 'Foo container'
        }
      })
      const extensions = [
        mock<Extension>({
          id: 'foo-1',
          userPreference: {
            optionLabel: 'Foo 1'
          }
        }),
        mock<Extension>({
          id: 'foo-2',
          userPreference: {
            optionLabel: 'Foo 2'
          }
        })
      ]
      const { wrapper } = getWrapper({
        extensionPoints: [extensionPoint],
        extensions
      })
      await blockLoadingState(wrapper)

      expect(wrapper.find(selectors.extensionsSection).exists()).toBeTruthy()
    })
  })
})

const blockLoadingState = async (wrapper: VueWrapper<any, any>) => {
  await wrapper.vm.loadAccountBundleTask.last
  await wrapper.vm.loadValuesListTask.last
  await wrapper.vm.loadGraphUserTask.last
}

function getWrapper({
  user = mock<User>({ memberOf: [] }),
  capabilities = {},
  accountEditLink = undefined,
  spaces = [],
  isPublicLinkContext = false,
  isUserContext = true,
  extensionPoints = [],
  extensions = []
}: {
  user?: User
  capabilities?: Partial<Capabilities['capabilities']>
  accountEditLink?: OptionsConfig['accountEditLink']
  spaces?: SpaceResource[]
  isPublicLinkContext?: boolean
  isUserContext?: boolean
  extensionPoints?: ExtensionPoint<Extension>[]
  extensions?: Extension[]
} = {}) {
  const plugins = defaultPlugins({
    piniaOptions: {
      userState: { user },
      authState: {
        userContextReady: isUserContext,
        publicLinkContextReady: isPublicLinkContext
      },
      spacesState: { spaces },
      capabilityState: { capabilities },
      configState: {
        options: {
          logoutUrl: 'https://account-manager/logout',
          ...(accountEditLink && { accountEditLink })
        }
      }
    }
  })

  const { getExtensionPoints, requestExtensions } = useExtensionRegistry()
  vi.mocked(getExtensionPoints).mockReturnValue(extensionPoints)
  vi.mocked(requestExtensions).mockReturnValue(extensions)

  const mocks = {
    ...defaultComponentMocks(),
    $route
  }

  mocks.$clientService.httpAuthenticated.post.mockImplementation((url) => {
    let response = {}

    if (url.endsWith('bundles-list')) {
      response = { bundles: [mock<SettingsBundle>()] }
    }
    if (url.endsWith('values-list')) {
      response = { values: [mock<SettingsValue>()] }
    }

    return Promise.resolve(mockAxiosResolve(response))
  })
  mocks.$clientService.graphAuthenticated.users.getMe.mockResolvedValue(mock<User>({ id: '1' }))

  return {
    mocks,
    wrapper: mount(account, {
      global: {
        plugins,
        mocks,
        provide: mocks,
        stubs: {
          'extension-preference': true
        }
      }
    })
  }
}
