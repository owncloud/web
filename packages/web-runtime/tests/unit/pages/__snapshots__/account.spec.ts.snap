// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`account page > account information section > displays basic user information 1`] = `
"<dl data-v-f2ac0579="" class="account-page-info oc-flex oc-flex-wrap">
  <div data-v-f2ac0579="" class="account-page-info-username oc-mb oc-width-1-2@m oc-width-1-1@s">
    <dt data-v-f2ac0579="" class="oc-text-normal oc-text-muted">Username</dt>
    <dd data-v-f2ac0579="">some-username</dd>
  </div>
  <div data-v-f2ac0579="" class="account-page-info-displayname oc-mb oc-width-1-2@m oc-width-1-1@s">
    <dt data-v-f2ac0579="" class="oc-text-normal oc-text-muted">First and last name</dt>
    <dd data-v-f2ac0579="">some-displayname</dd>
  </div>
  <div data-v-f2ac0579="" class="account-page-info-email oc-mb oc-width-1-2@m oc-width-1-1@s">
    <dt data-v-f2ac0579="" class="oc-text-normal oc-text-muted">Email</dt>
    <dd data-v-f2ac0579="">some-email</dd>
  </div>
  <div data-v-f2ac0579="" class="account-page-info-groups oc-mb oc-width-1-2@m oc-width-1-1@s">
    <dt data-v-f2ac0579="" class="oc-text-normal oc-text-muted">Group memberships</dt>
    <dd data-v-f2ac0579="" data-testid="group-names"><span data-v-f2ac0579="" data-testid="group-names-empty">You are not part of any group</span></dd>
  </div>
  <div data-v-f2ac0579="" class="account-page-logout-all-devices oc-mb oc-width-1-2@m oc-width-1-1@s">
    <dt data-v-f2ac0579="" class="oc-text-normal oc-text-muted">Logout from active devices</dt>
    <dd data-v-f2ac0579="" data-testid="logout">
      <oc-button-stub data-v-f2ac0579="" appearance="raw" type="a" href="https://account-manager/logout" target="_blank" data-testid="account-page-logout-url-btn"></oc-button-stub>
    </dd>
  </div>
  <!--v-if-->
</dl>"
`;

exports[`account page > account information section > group membership > displays group names 1`] = `"<dd data-v-f2ac0579="" data-testid="group-names"><span data-v-f2ac0579="">one, two, three</span></dd>"`;

exports[`account page > header section > edit url button > should be displayed if defined via config 1`] = `"<oc-button-stub data-v-f2ac0579="" variation="primary" type="a" href="/" target="_blank" data-testid="account-page-edit-url-btn"></oc-button-stub>"`;

exports[`account page > public link context > should render a limited view 1`] = `
"<main data-v-f2ac0579="" id="account" class="oc-height-1-1 oc-m">
  <h1 data-v-f2ac0579="" id="account-page-title" class="oc-page-title oc-m-rm oc-invisible-sr">Some Title</h1>
  <!--v-if-->
  <div data-v-f2ac0579="">
    <div data-v-f2ac0579="" class="oc-flex oc-width-1-1">
      <h2 data-v-f2ac0579="" class="oc-text-bold">Preferences</h2>
    </div>
    <dl data-v-f2ac0579="" class="account-page-preferences oc-flex oc-flex-wrap">
      <div data-v-f2ac0579="" class="account-page-info-language oc-mb oc-width-1-2@m oc-width-1-1@s">
        <dt data-v-f2ac0579="" class="oc-text-normal oc-text-muted">Language</dt>
        <dd data-v-f2ac0579="" data-testid="language" class="oc-width-1-3@l oc-width-1-2@m oc-width-1-1@s">
          <oc-select-stub data-v-f2ac0579="" id="oc-select-1" filter="[Function]" disabled="false" optionlabel="label" getoptionlabel="[Function]" searchable="true" clearable="false" loading="false" fixmessageline="false" multiple="false" readonly="false" positionfixed="false" placeholder="Please choose..." options="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]" model-value="[object Object]"></oc-select-stub>
          <div data-v-f2ac0579="" class="oc-mt-s"><a data-v-f2ac0579="" href="https://explore.transifex.com/owncloud-org/owncloud-web/" target="_blank" class="oc-flex oc-flex-middle">
              <oc-icon-stub data-v-f2ac0579="" name="service" class="oc-mr-xs"></oc-icon-stub> <span data-v-f2ac0579="">Help translate</span>
            </a></div>
        </dd>
      </div>
      <div data-v-f2ac0579="" class="account-page-info-theme oc-mb oc-width-1-2@m oc-width-1-1@s">
        <dt data-v-f2ac0579="" class="oc-text-normal oc-text-muted">Theme</dt>
        <dd data-v-f2ac0579="" data-testid="theme" class="oc-width-1-3@l oc-width-1-2@m oc-width-1-1@s">
          <theme-switcher-stub data-v-f2ac0579=""></theme-switcher-stub>
        </dd>
      </div>
      <!--v-if-->
      <!--v-if-->
    </dl>
  </div>
  <!--v-if-->
  <!--v-if-->
</main>"
`;
